devtools/onion.c

```diff
@@ -12,32 +13,55 @@
 #include <unistd.h>
 
 #define ASSOC_DATA_SIZE 32
+#define PUBKEY_LEN 33
+#define PRIVKEY_LEN 32
 
 static void do_generate(int argc, char **argv,
 			const u8 assocdata[ASSOC_DATA_SIZE])
 {
 	const tal_t *ctx = talz(NULL, tal_t);
 	int num_hops = argc - 1;
 	struct pubkey *path = tal_arr(ctx, struct pubkey, num_hops);
-	u8 privkeys[argc - 1][32];
+	u8 rawpubkey[PUBKEY_LEN], rawprivkey[PRIVKEY_LEN];
 	struct secret session_key;
 	struct hop_data hops_data[num_hops];
 	struct secret *shared_secrets;
 	struct sphinx_path *sp;
 
-	assocdata = tal_arr(ctx, u8, 32);
+	assocdata = tal_arr(ctx, u8, ASSOC_DATA_SIZE);
 	memset(&session_key, 'A', sizeof(struct secret));
 
 	sp = sphinx_path_new_with_key(ctx, assocdata, &session_key);
 
 	for (int i = 0; i < num_hops; i++) {
 		size_t klen = strcspn(argv[1 + i], "/");
-		if (!hex_decode(argv[1 + i], klen, privkeys[i], 32))
-			errx(1, "Invalid private key hex '%s'", argv[1 + i]);
-
-		if (secp256k1_ec_pubkey_create(secp256k1_ctx, &path[i].pubkey,
-					       privkeys[i]) != 1)
-			errx(1, "Could not decode pubkey");
+		assert(klen == 2 * PUBKEY_LEN || klen == 2 * PRIVKEY_LEN);
+		if (klen == 2 * PRIVKEY_LEN) {
+			if (!hex_decode(argv[1 + i], klen, rawprivkey, PRIVKEY_LEN))
```

Comment:
I would combine these two, for simplicity.  hex_decode() will fail if the wrong length, and it avoids magic `2 *`.  The other option would be `if (hex_data_size(klen) == PRIVKEY_LEN)` but it still seems redundant?

Reply:
Switched to use `hex_data_size` and `pubkey_from_hexstr` to do the conversion in one go. Would a counterpart for privkeys be desirable as well (`privkey_from_hexstr`)?

Reply:
Hmm, hex_decode *does* this test internally, so it's redundant.  But not harmful, and I really want this merged :)